File: server/test/validate.test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');
const { isValidImageRef } = require('../src/validate');

test('accepts plausible image / OCI artifact references', () => {
  assert.equal(isValidImageRef('alpine:latest'), true);
  assert.equal(isValidImageRef('docker.io/library/nginx:1.27'), true);
  assert.equal(isValidImageRef('registry.example.com:5000/team/app@sha256:' + 'a'.repeat(64)), true);
  assert.equal(isValidImageRef('ghcr.io/org/artifact:oci-1.0.0'), true);
});

test('rejects empty, non-string or malformed references', () => {
  assert.equal(isValidImageRef(''), false);
  assert.equal(isValidImageRef('   '), false);
  assert.equal(isValidImageRef(undefined), false);
  assert.equal(isValidImageRef(null), false);
  assert.equal(isValidImageRef(42), false);
  assert.equal(isValidImageRef(' alpine:latest'), false);
  assert.equal(isValidImageRef('alpine latest'), false);
  assert.equal(isValidImageRef('image;rm -rf /'), false);
  assert.equal(isValidImageRef('$(whoami)'), false);
  assert.equal(isValidImageRef('`id`'), false);
});

File: pkg/validate/validate_test.go
package validate

import "testing"

func TestIsValidImageRefAccepts(t *testing.T) {
	valid := []string{
		"alpine:latest",
		"docker.io/library/nginx:1.27",
		"registry.example.com:5000/team/app@sha256:" + repeat("a", 64),
		"ghcr.io/org/artifact:oci-1.0.0",
	}
	for _, image := range valid {
		if !IsValidImageRef(image) {
			t.Errorf("expected %q to be valid", image)
		}
	}
}

func TestIsValidImageRefRejects(t *testing.T) {
	invalid := []string{
		"",
		"   ",
		" alpine:latest",
		"alpine latest",
		"image;rm -rf /",
		"$(whoami)",
		"`id`",
	}
	for _, image := range invalid {
		if IsValidImageRef(image) {
			t.Errorf("expected %q to be invalid", image)
		}
	}
}

func repeat(s string, n int) string {
	out := make([]byte, 0, n*len(s))
	for i := 0; i < n; i++ {
		out = append(out, s...)
	}
	return string(out)
}
